File: python/src/semantic_retrieval/prompts/prompt_template.py
from typing import Dict, Optional

from semantic_retrieval.prompts.prompt import IPrompt

PromptTemplateParameters = Dict[str, str]


class PromptTemplate(IPrompt):
    def __init__(self, template: str, params: PromptTemplateParameters):
        self.template = template
        self.parameters = params

    def set_parameters(self, params: PromptTemplateParameters):
        self.parameters = params

    def resolve_template(
        self, parameters: Optional[PromptTemplateParameters] = None
    ):
        return ""

    async def to_string(self) -> str:
        return self.resolve_template()

File: typescript/src/prompts/prompt-templates/promptTemplate.ts
import { compile } from "handlebars";
import { IPrompt } from "../prompt";

export type PromptTemplateParameters = { [key: string]: string };
/**
 * A class for defining reusable prompts with template strings. This base class
 * uses handlebars for template resolution by default.
 */
export class PromptTemplate implements IPrompt {
  template: string;
  parameters: PromptTemplateParameters = {};

  constructor(template: string, params?: PromptTemplateParameters) {
    this.template = template;
    this.parameters = params ?? this.parameters;
  }

  setParameters(params: PromptTemplateParameters) {
    this.parameters = params;
  }

  resolveTemplate(parameters?: PromptTemplateParameters): string {
    const compiledTemplate = compile(this.template);
    return compiledTemplate(parameters ?? this.parameters);
  }

  async toString(): Promise<string> {
    return this.resolveTemplate();
  }
}
